sycl/source/detail/kernel_impl.cpp

```diff
@@ -21,8 +23,48 @@ kernel_impl::kernel_impl(RT::PiKernel Kernel, const context &SyclContext)
                   std::make_shared<program_impl>(SyclContext, Kernel),
                   /*IsCreatedFromSource*/ true) {}
 
+kernel_impl::kernel_impl(RT::PiKernel Kernel, const context &SyclContext,
+                         std::shared_ptr<program_impl> ProgramImpl,
+                         bool IsCreatedFromSource)
+    : MKernel(Kernel), MContext(SyclContext),
+      MProgramImpl(std::move(ProgramImpl)),
+      MCreatedFromSource(IsCreatedFromSource) {
+
+  RT::PiContext Context = nullptr;
+  PI_CALL(RT::piKernelGetInfo, MKernel, CL_KERNEL_CONTEXT, sizeof(Context),
+          &Context, nullptr);
+  auto ContextImpl = detail::getSyclObjImpl(SyclContext);
+  if (ContextImpl->getHandleRef() != Context)
+    throw cl::sycl::invalid_parameter_error(
+        "Input context must be the same as the context of cl_kernel");
+  PI_CALL(RT::piKernelRetain, MKernel);
+}
+
+kernel_impl::kernel_impl(const context &SyclContext,
+                         std::shared_ptr<program_impl> ProgramImpl)
+    : MContext(SyclContext), MProgramImpl(std::move(ProgramImpl)) {}
+
+kernel_impl::~kernel_impl() {
+  // TODO catch an exception and put it to list of asynchronous exceptions
```

Comment:
@romanovvlad, isn't this synchronous operation?

Reply:
What do you mean? Destructor cannot be synchronous or asynchronous by itself.
Anyway, the idea is to avoid throwing exceptions from destructors and put them to the list of
asynchronous exceptions to still let user know something wrong happened.

Reply:
> What do you mean? Destructor cannot be synchronous or asynchronous by itself.

I'm talking about `clReleaseKernel` call.

> Anyway, the idea is to avoid throwing exceptions from destructors and put them to the list of
> asynchronous exceptions to still let user know something wrong happened.

I don't think it's asynchronous exception is the right mechanism to report errors occurred in SYCL object destructors for non-asynchronous operations.

Should we just remove this comment?

Reply:
I'm talking about clReleaseKernel call is synchronous. It decrements reference counter during the call.

> I don't think it's asynchronous exception is the right mechanism to report errors occurred in SYCL object destructors for non-asynchronous operations.

> Should we just remove this comment?

Before removing, do you have better ideas on how to report such errors?

Reply:
assert?
I don't this correctly implemented `kernel_impl` can cause any errors here.

Reply:
Consider the following code:
{
  sycl::kernel Kenrel(); <= cl_kernel ref count == 1
  cl_kernel CLKernel = Kernel.get(); <= cl_kernel ref count == 2
  clReleaseKernel(CLKernel); <= cl_kernel ref count == 1
  clReleaseKernel(CLKernel); <= cl_kernel ref count == 0
}  <= clReleaseKernel cann in sycl::kernel d'tor fails.

Reply:
Who said we always should call a `clReleaseKernel` here?
It doesn't make sense if reference counter is 0.

Reply:
Ok, to be more specific, could you please suggest implementation of kernel_impl destructor which is guaranteed to work errors? kernel_impl should decrement cl_kernel reference counter otherwise it is memory leak.